### PR TITLE
Recommend Terraform over CloudFormation

### LINKED ISSUE #6: We should trial Terraform

### PR DESCRIPTION
**Pull request by rlmartin**
Here is an example of what a recommendation PR might look like.

Fixes #6.

### CODE REVIEW

**Comment by moniecodes** [comment]
Ok I see. Not sure I totally get it all. Can these lists be integrated somewhere else? Like pulled into our “intranet” or a a dashboard or widget component? Are we basically generating a standards wiki?

**Comment by figgybit** [comment]
following up on the terraform vs. cloudformation conversations.  cloudformation is quite easy and fast to develop.  In addition if you are not expecting to use anything more complex than lambdas, sns, sqs, and ddb then it is easy to implement.  I am curious about taking on a project to revamp our usage of serverless.com's yaml into terraform.  However at this time, perhaps that is not something that would warrant our focus and energy.  I am up for us to explore using terraform for future development.

File: recommendations/AWS.md [hunk 1]
```diff
@@ -2,3 +2,53 @@
...
+2. Underlying configuration that is produced (CloudFormation, Terraform, etc) should be commited to a repository.
+3. Keep cost in mind.
+
+In the meantime, we should continue working with established tools and follow best-practices.
+
+
+## Trial
+
+### Terraform
+Moda should be capturing infrastructure assets in committed code ([why it is important](https://www.hashicorp.com/resources/what-is-infrastructure-as-code)). Current thinking is that it is superior to [CloudFormation](#cloudformation) for the following reasons:
```
Comment (celliott) [inline recommendations/AWS.md]: 3. Terraform is infrastructure as code where CloudFormation is large `yaml` or `json` definitions
4. Terraform is much easier to read and write than CloudFormation
5. We can write Terraform modules to increase infrastructure reproducibility, consistency, and iteration
6. Terraform is easier to learn than CloudFormation
7. Hashicorp is actively developing terraform
8. Active open source community. Lots of contributors, docs, modules, examples, videos, talks, POCs, etc
9. Engineers are already using and learning Terraform at Moda
10. Arguably a better fit than CloudFormation for code generation and iteration

Comment (rlmartin) [reply recommendations/AWS.md]: I added these to the doc.

**Comment by rlmartin** [comment]
Consensus from the group that this is the path forward.

File: recommendations/AWS.md [hunk 2]
```diff
@@ -2,3 +2,60 @@
...
+
+### CloudFormation
+Moda should be capturing infrastructure assets in committed code ([why it is important](https://www.hashicorp.com/resources/what-is-infrastructure-as-code)). CloudFormation is the AWS-native way to do IaC. Please see commentary on [Terraform](#terraform) for reasons why CloudFormation is not recommended.
+
+#### Note regarding AWS CDK (Cloud Development Kit)
+Per the [CDK FAQ](https://aws.amazon.com/cdk/faqs):
+```
+Since AWS CDK leverages CloudFormation, AWS CDK applications are subject to the same limits imposed by the CloudFormation service.
+```
+As such, the CDK should be considered on `Hold` because CloudFormation itself is.
```
Comment (barryoneill) [inline recommendations/AWS.md]: I would argue that the CDK, now in 1.0.0 GA should be in assess.  Sure, it's subject under the hood to the same 'limits', but I think we shouldn't be so quick to dismiss the power that high level languages (as offered by CDK, Pulumi and co) bring and how they can be used to work around some of the really awful hacks that I see in terraform for looping and dynamic infrastructure definitions.   

I understand this discussion is closed for now, but I'd encourage people to keep track of how much weight AWS throw behind the CDK, and keep an open mind towards it (and its ilk, i.e. Pulumi)

File: recommendations/AWS.md [hunk 3]
```diff
@@ -2,3 +2,60 @@
...
+
+## Trial
+
+### Terraform
+Moda should be capturing infrastructure assets in committed code ([why it is important](https://www.hashicorp.com/resources/what-is-infrastructure-as-code)). Current thinking is that it is superior to [CloudFormation](#cloudformation) for the following reasons:
+
+  1. Terraform is multi-service. This does not mean that the same Terraform resource can be deployed indiscriminately across different clouds. But it does mean that resources across multiple cloud providers (or any provider, really) can be captured in Terraform - and thus be captured in the same workflow, pipelines, etc.
+  2. Terraform uses the provider's APIs. As a result, Terraform - which can model anything available via the API - will always be ahead of CloudFormation in supported resources.
+  3. Terraform is infrastructure as code where CloudFormation is large yaml or json definitions.
+  4. Terraform is much easier to read and write than CloudFormation.
```
Comment (barryoneill) [inline recommendations/AWS.md]: But still a lot less intuitive and easier to read than a high level language.

File: recommendations/AWS.md [hunk 4]
```diff
@@ -2,3 +2,60 @@
...
+  5. We can write Terraform modules to increase infrastructure reproducibility, consistency, and iteration.
+  6. Terraform is easier to learn than CloudFormation.
```
Comment (barryoneill) [inline recommendations/AWS.md]: But is still another DSL to learn

File: recommendations/AWS.md [hunk 5]
```diff
@@ -2,3 +2,60 @@
...
+  7. Hashicorp is actively developing terraform.
+  8. Active open source community. Lots of contributors, docs, modules, examples, videos, talks, POCs, etc.
```
Comment (barryoneill) [inline recommendations/AWS.md]: Except in my experience, for the first while after AWS releases new functionality, if you use those services, prepare to be contributing to the AWS providers yourself as well.

**Pull request merged**